File: Kickstart/2021/Round F/B. Festival/2021k-f-b.py
# 2021 Kickstart Round F - B. Festival
# https://codingcompetitions.withgoogle.com/kickstart/round/0000000000435bae/0000000000887dba

from collections import defaultdict, namedtuple
from bisect import insort_left

Int = lambda: int(input())
Ints = lambda: [*map(int, input().split())]
Lines = lambda line, n_lines: [line() for _ in range(n_lines)]

Attraction = namedtuple('Attraction', 'happy, start, end')

#------------------------------------------------------------------------------#

def solve_small(d, n, k, attractions):
    days = defaultdict(list)
    for a in attractions:
        a = a._replace(end = a.end+1)
        days[a.start].append(a)
        days[a.end].append(a)

    def start(a):
        nonlocal curr
        if len(actives) < k:
            insort_left(actives, a)
            curr += a.happy
        elif a.happy >= actives[0].happy:
            sub = actives.pop(0)
            insort_left(subs, sub)
            insort_left(actives, a)
            curr -= sub.happy
            curr += a.happy
        else:
            insort_left(subs, a)

    def end(a):
        nonlocal curr
        if a in subs: 
            subs.remove(a)
        if a in actives: 
            actives.remove(a)
            curr -= a.happy
            if subs: 
                sub = subs.pop()
                insort_left(actives, sub)
                curr += sub.happy
            
    actives, subs = [], []
    ans = curr = 0
    for day in sorted(days):
        for a in days[day]:
            if day == a.end: end(a)
            if day == a.start: start(a)
        ans = max(ans, curr)
    return ans

#------------------------------------------------------------------------------#

for i in range(Int()):
    d, n, k = Ints()
    attractions = Lines(lambda: Attraction(*Ints()), n)
    result = solve_small(d, n, k, attractions)
    print('Case #{}:'.format(i+1), result)

File: Kickstart/2021/Round F/B. Festival/2021k-f-b.cpp
// 2021 Kickstart Round F - B. Festival
// https://codingcompetitions.withgoogle.com/kickstart/round/0000000000435bae/0000000000887dba

#include <bits/stdc++.h>
using namespace std;

#define rep(i, j, k) for (int i=j; i<=k; i++)
#define For(stop) rep(i, 0, stop-1)
#define pb push_back
using ll = long long;
using vi = vector<int>;

/*----------------------------------------------------------------------------*/ 
 
array<vi, 300001> ops, eds;
 
ll solve() {
    int D, N, K; cin >> D >> N >> K;

    vi H(N); 
    For(N) {
        int S, E; 
        cin >> H[i] >> S >> E;
        ops[--S].pb(i); eds[E].pb(i);
    }

    ll ans(0), curr(0);
    multiset<int> actives, subs;
    For(D+1) {
        for (int d: ops[i]) {
            actives.insert(H[d]);
            curr += H[d];
            if (actives.size() > K) {
                auto it = begin(actives);
                curr -= *it;
                subs.insert(*it);
                actives.erase(it);
            }
        }
        ops[i].clear();

        for (int d: eds[i]) {
            if (subs.find(H[d]) != end(subs)) {
                subs.erase( subs.find(H[d]) );
            } else {
                actives.erase( actives.find(H[d]) );
                curr -= H[d];
                if (subs.size()) {
                    auto it = --subs.end();
                    curr += *it;
                    actives.insert(*it);
                    subs.erase(it);
                }
            }
        }
        eds[i].clear();

        ans = max(ans, curr);
    }
    return ans;
}

/*----------------------------------------------------------------------------*/

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int T; cin >> T;
    For(T) cout << "Case #" << i+1 << ": " << solve() << endl;
}